programs/perpetuals/tests/instructions/test_add_custody.rs

```diff
@@ -0,0 +1,120 @@
+use crate::utils::{self, pda};
+use anchor_lang::{
+    prelude::{AccountMeta, Pubkey},
+    InstructionData, ToAccountMetas,
+};
+use perpetuals::{
+    instructions::AddCustodyParams,
+    state::{custody::Custody, multisig::Multisig, pool::Pool},
+};
+use solana_program_test::ProgramTestContext;
+use solana_sdk::signer::{keypair::Keypair, Signer};
+
+pub async fn test_add_custody(
+    program_test_ctx: &mut ProgramTestContext,
+    admin: &Keypair,
+    payer: &Keypair,
+    pool_pda: &Pubkey,
+    custody_token_mint: &Pubkey,
+    custody_token_decimals: u8,
+    params: AddCustodyParams,
+    multisig_signers: &[&Keypair],
+) -> (anchor_lang::prelude::Pubkey, u8) {
+    // ==== WHEN ==============================================================
+    let multisig_pda = pda::get_multisig_pda().0;
+    let transfer_authority_pda = pda::get_transfer_authority_pda().0;
+    let perpetuals_pda = pda::get_perpetuals_pda().0;
+    let (custody_pda, custody_bump) = pda::get_custody_pda(pool_pda, custody_token_mint);
+    let (custody_token_account_pda, custody_token_account_bump) =
+        pda::get_custody_token_account_pda(pool_pda, custody_token_mint);
+
+    let multisig_account = utils::get_account::<Multisig>(program_test_ctx, multisig_pda).await;
+
+    // One Tx per multisig signer
+    for i in 0..multisig_account.min_signatures {
+        let signer: &Keypair = multisig_signers[i as usize];
+
+        let accounts_meta = {
+            let accounts = perpetuals::accounts::AddCustody {
+                admin: admin.pubkey(),
+                multisig: multisig_pda,
+                transfer_authority: transfer_authority_pda,
+                perpetuals: perpetuals_pda,
+                pool: *pool_pda,
+                custody: custody_pda,
+                custody_token_account: custody_token_account_pda,
+                custody_token_mint: *custody_token_mint,
+                system_program: anchor_lang::system_program::ID,
+                token_program: anchor_spl::token::ID,
+                rent: solana_program::sysvar::rent::ID,
+            };
+
+            let mut accounts_meta = accounts.to_account_metas(None);
+
+            accounts_meta.push(AccountMeta {
+                pubkey: signer.pubkey(),
+                is_signer: true,
+                is_writable: false,
+            });
+
+            accounts_meta
+        };
+
+        let arguments = perpetuals::instruction::AddCustody { params };
+
+        let ix = solana_sdk::instruction::Instruction {
```

Comment:
have you thought of creating a wrapper for the code below so you don't need to repeat it for every instruction tested? Also instead of unwrapping to handle and print the error more gracefully?